apps/labs/posts/numpy-quaddtype-blog.md

```diff
@@ -0,0 +1,419 @@
+
+```c
+typedef union {
+    Sleef_quad sleef_value;
+    long double longdouble_value;
+} quad_value;
+
+typedef struct {
+    PyObject_HEAD
+    quad_value value;
+    QuadBackendType backend;
+} QuadPrecisionObject;
+```
+
+This dual-nature allows us to leverage the full 128-bit precision of SLEEF when available, while maintaining backwards compatibility with `np.longdouble`.
+
+To integrate seamlessly with NumPy, we've created the `QuadPrecDTypeObject`:
+
+```c
+typedef struct {
+    PyArray_Descr base;
+    QuadBackendType backend;
+} QuadPrecDTypeObject;
+```
+
+This structure acts as a bridge, allowing our high-precision numbers to work harmoniously within NumPy arrays and operations.
+
+The backend selection is handled by a simple enum:
+
+```c
+typedef enum {
+    BACKEND_INVALID = -1,
+    BACKEND_SLEEF,
+    BACKEND_LONGDOUBLE
+} QuadBackendType;
+```
+
+This flexibility enables `numpy_quaddtype` to provide optimal precision across different platforms while maintaining a consistent interface. Users can select their preferred `backend` at runtime:
+
+```python
+import numpy as np
+import numpy_quaddtype as npq
+
+# Using SLEEF backend (default)
+x = npq.QuadPrecision(3.14159)
+x =npq.QuadPrecision(3.14159, backend='sleef')
+
+# Using longdouble backend
+y = npq.(2.71828, backend='longdouble')
+
+# Creating a NumPy array with QuadPrecision dtype
+z = np.array([x, y], dtype=npq.QuadPrecDType()) # SLEEF
+z = np.array([x, y], dtype=npq.QuadPrecDType("longdouble")) # SLEEF
+```
+
+Under the hood, `numpy_quaddtype` manages memory efficiently for both **aligned** and **unaligned**. This is crucial for performance, especially when dealing with large arrays or complex computations. We've implemented specialized strided loop functions for various operations
+
+<figure style={{ textAlign: 'center' }}>
+    <img 
+      src="/posts/numpy-quaddtype-blog/aligned_unaligned_memory.png"
+      alt="Diagram illustrating an example of aligned versus unaligned memory"
+      style={{ display: 'inline-block', maxWidth: '100%', height: 'auto' }}
+    />
+    <figcaption>Figure 1: Illustrating an example of aligned and unaligned memory</figcaption>
+</figure>
+
+
+
+### Casting operations
+Casting operations are another critical component. We've implemented a range of casts between QuadPrecision and other NumPy types, ensuring smooth interoperability:
+
+```python
+import numpy as np
+import numpy_quaddtype as npq
+
+# NumPy to QuadPrecision
+numpy_array = np.array([1.5, 2.7, 3.14])
+quad_array = numpy_array.astype(npq.QuadPrecDType())
+
+# QuadPrecision to NumPy
+quad_value = QuadPrecision("3.14159265358979323846")
+numpy_float = np.float64(quad_value)
+
+# Mixing types in operations
+result = quad_array + numpy_array  # Automatically promotes to QuadPrecision
```

Comment:
If you write this where the input lines are prepended with `>>>` you can then include the output of the operation just like if this was a copy/paste of a terminal session. It's easier to understand with the inline output, I think.